clap_derive/src/derives/from_arg_matches.rs

```diff
@@ -239,3 +246,198 @@ pub fn gen_constructor(fields: &Punctuated<Field, Comma>, parent_attribute: &Att
         #( #fields ),*
     }}
 }
+
+pub fn gen_updater(
+    fields: &Punctuated<Field, Comma>,
+    parent_attribute: &Attrs,
+    use_self: bool,
+) -> TokenStream {
+    let fields = fields.iter().map(|field| {
+        let attrs = Attrs::from_field(
+            field,
+            parent_attribute.casing(),
+            parent_attribute.env_casing(),
+        );
+        let field_name = field.ident.as_ref().unwrap();
+        let kind = attrs.kind();
+
+        let access = if use_self {
+            quote!{
+                #[allow(non_snake_case)]
+                let #field_name = &mut self.#field_name; }
+        } else {
+            quote!()
+        };
+
+        match &*kind {
+            Kind::ExternalSubcommand => {
+                abort! { kind.span(),
+                    "`external_subcommand` can be used only on enum variants"
+                }
+            }
+            Kind::Subcommand(ty) => {
+                let subcmd_type = match (**ty, sub_type(&field.ty)) {
+                    (Ty::Option, Some(sub_type)) => sub_type,
+                    _ => &field.ty,
+                };
+
+                let updater = quote_spanned!{ ty.span()=>
+                    <#subcmd_type as ::clap::Subcommand>::update_from_subcommand(#field_name, subcmd);
+                };
+
+                let updater = match **ty {
+                    Ty::Option => quote_spanned! { kind.span()=>
+                        if let Some(#field_name) = #field_name.as_mut() {
+                            #updater
+                        } else {
+                            *#field_name = <#subcmd_type as ::clap::Subcommand>::from_subcommand(
+                                subcmd
+                            )
+                        }
+                    },
+                    _ => quote_spanned!{ kind.span()=>
+                        #updater
+                    }
+                };
+
+                quote_spanned! { kind.span()=>
+                    {
+                        let subcmd = matches.subcommand();
+                        #access
+                        #updater
+                    }
+                }
+            }
+
+            Kind::Flatten => quote_spanned! { kind.span()=> {
+                    #access
+                    ::clap::FromArgMatches::update_from_arg_matches(#field_name, matches);
+                }
+            },
+
+            Kind::Skip(_) => quote!(),
+
+            Kind::Arg(ty) => {
+                use self::ParserKind::*;
+
+                let parser = attrs.parser();
+                let func = &parser.func;
+                let span = parser.kind.span();
+                let (value_of, values_of, mut parse) = match *parser.kind {
```

Comment:
Can't we extract this since it looks to be common instead of repeating?

Reply:
there is a small difference, let me see if I can factor out that code and then account for it.